marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/InputOutputEndpointImpl.java

```diff
@@ -135,11 +204,43 @@ private void processInput() {
 
         @Override
         public void awaitCompletion() {
-            if (getQueue() == null)
-                return;
+            try {
+                if (getQueue() == null)
+                    return;
+
+                while (getQueue() != null && !getQueue().isEmpty()) {
+                    processInput();
+                }
+                getCallerThreadPoolExecutor().awaitTermination();
+            } catch (Throwable throwable) {
+                throw new RuntimeException("Error occurred while awaiting termination "+throwable.getMessage());
+            }
+        }
 
-            while (!getQueue().isEmpty()) {
-                processInput();
+        private class BulkCallableImpl implements Callable<Boolean> {
+            private BulkInputOutputCallerImpl bulkInputOutputCallerImpl;
+            private InputStream[] inputBatch;
+            BulkCallableImpl(BulkInputOutputCallerImpl BulkInputOutputCallerImpl, InputStream[] inputBatch) {
+                this.bulkInputOutputCallerImpl = BulkInputOutputCallerImpl;
+                this.inputBatch = inputBatch;
+            }
+            @Override
+            public Boolean call() {
+                try {
+                    CallContext callContext = bulkInputOutputCallerImpl.getCallContextQueue().poll();
+                    if(callContext != null)
+                        bulkInputOutputCallerImpl.processInput(callContext, inputBatch);
+
+                    if(getPhase() == WorkPhase.COMPLETED && bulkInputOutputCallerImpl.getCallContextQueue().isEmpty()) {
+                        getCallerThreadPoolExecutor().shutdown();
+                    }
+                    else if(callContext != null) {
```

Comment:
Here also, it is important to check for a null call context before checking for shutdown.